### PR TITLE
New command: spack cmake - Configure a CMake project using a Spack spec 

### PR DESCRIPTION
**Pull request by rbberger**
@tgamblin Here is my attempt to upstream the tool I mentioned in Slack. I've modified it a bit to be more like a regular Spack command, reworked some of the output, added a `--dry-run` option, and stripped some project specific features. The code can likely still be generalized a bit more. Right now, it is very specific to C/C++-based projects that use Kokkos for HIP and CUDA. Below is a bit of motivating text, part of which might end up in the final documentation.

Let me know what you think and how it could be improved to better fit into Spack.

---

This command was built out of a need to simplify developer's workflows while working on codes that also provide Spack packages. The original  [`spack_cmake`](https://github.com/flecsi/flecsi/blob/2.3/tools/spack_cmake) tool was written for the development and CI testing of FleCSI.

We recognized that many complicated use cases and the corresponding CMake command-lines were already
encoded in the Spack package for a given code, and we needed a way to extract that useful logic in a pure CMake-based workflow and avoid duplication.

This new command is closely related to `spack develop` and `spack dev-build`. However, unlike these two existing commands the goal of `spack cmake` is not to build and install the package within Spack, but allow it to be configured and built outside of the Spack build environment and within the same shell (no sub-shells). This is particularly useful in non-interactive CI pipelines. It assumes package dependencies are already visible either by an active environment view, via `spack load`, or modules.

Its primary purpose is to reuse the logic in the Spack package of a given project and generate an equivalent CMake command-line out of a given root spec. Instead of using Spack compiler wrappers, it will inject the compiler's full paths
via the `CMAKE_<LANG>_COMPILER` options.

While the Spack build environment and the Spack compiler wrappers do a lot of heavy-lifting, such as injecting compiler and linker flags, `spack cmake` so far does not attempt to provide feature-parity and focuses on a small subset that has been found useful.

In addition to the generated options that serve as defaults, the command allows to append additional CMake options to the command-line. Options passed after a double-dash (`--`) are directly passed along. (For this the Spack `main.py` needed to be changed to allow this kind of unparsed command-line options)

If `spack cmake` is run in an activated environment, the input spec is compared to that environment and any incompatible differences in its dependency tree are reported as warnings. If no compiler was specified, the spec defaults to using the same compiler as the root spec in that environment. With the `-i/--inherit` option the first compatible concretized root of the environment is selected, otherwise a fresh concretization happens.

By default, this command works out-of-the box if you run it from within the source directory of a CMake project or from a build directory that is a child of the source directory.

 - When launched from the source directory, it will create a `build` folder and configure it.

 - When launched from any other build directory that is a child of the source directory, it will use that build directory and locate the source folder.

 - For any other launch location, the current directory is assumed to be the build directory, unless it is overridden using the `-B <path>` option, and you  must specify the source directory with `-S <path>`

The command will execute the CMake command-line, unless the `-d,--dry-run` flag is set.

Some packages have special logic for testing that depends on the package variable `run_tests`.  The `-t/--test` option sets `run_tests` to `True`.

Finally, the `-c/--compiler-only` option allows to suppress project-specific CMake options. It will only include the selection of the compiler, compiler flags, and user specified options. This is useful for building dependent projects that require the same compiler settings as a given spec (some of which might have been manipulated by the `cmake_args` logic in a Spack spackage).

### Inherit root spec from environment to generate CMake command-line
```bash
spack env create myenv
spack env activate myenv
spack add flecsi%gcc+flog
spack install --only dependencies

spack cmake -i -t flecsi
# inherit flecsi spec from environment, enable testing
```

### Reuse existing environment with different options

```bash
spack cmake flecsi~flog build_type=Debug
# new concretization without flog and Debug as build type. In this case the dependency tree is the
# same, so the existing environment can be reused for this CMake build
```

### Pass additional CMake options

```bash
spack cmake -i flecsi -- -DCMAKE_INSTALL_PREFIX=$PWD/test-install
# reuse environment spec for flecsi and append CMake options
```

### Use same compiler settings as a given root

```bash
spack cmake -i -c flecsi+rocm
# useful in the case a standalone application needs to use the same compiler as the library,
# but doesn't need the library specific CMake options.
```

### CODE REVIEW

**Comment by psakievich** [comment]
TLDR
---
Overall I like many parts of what this feature is aiming to accomplish. At first blush it still seems very targeted with several assumptions that users must keep in mind to use effectively.  

I would also be interested in exploring what the driving deficiencies are in `spack develop` and `spack dev-build` in greater detail to see if the existing commands can be improved to meet these needs and do one of the following: 

a) reduce the specificity and assumptions required here
b) remove the need for a specific cmake command altogether

Some specific thoughts on the summary below:
---

> We recognized that many complicated use cases and the corresponding CMake command-lines were already
encoded in the Spack package for a given code, and we needed a way to extract that useful logic in a pure CMake-based workflow and avoid duplication.

I have received requests for something similar. Something like a `spack cmake export` feature to allow for reproducers for other projects.  An example would be an application using `trilinos` would like to send the trilinos developers a reproducer from their custom trilinos spack package without having to export their entire build system to the trilinos team.

> It assumes package dependencies are already visible either by an active environment view, via spack load, or modules.

It would be really nice to remove this as an assumption and make it a property of the command.  For example, since we are assuming manipulation of the current shell, is there a compelling reason to not do `source $(spack build-env [package])` in the current shell?  The only reason we launch subshells for the current commands is to keep the active shell unmodified. This could be reconfigured with an option to help with this CI use case without adding a whole new command/interface.

We could also look into exporting an artifact so this environment can be recreated over and over again.

> This new command is closely related to spack develop and spack dev-build. However, unlike these two existing commands the goal of spack cmake is not to build and install the package within Spack, but allow it to be configured and built outside of the Spack build environment and within the same shell (no sub-shells)

Same notes as above, but also makes me wonder what else don't you like about the current tools?  This command seems like one of the driving reasons is to move out of the current `spack` tools and back to what others are more used to. While there is merit to that, it would be helpful to understand what the deficiencies are since the spack tooling is generalized enough to support more build systems than just CMake and working on more than one application at at time.

My first introduction to spack was using `spack location -i` inside a cmake script to configure a root application which this seems like an automated version of. 

> In addition to the generated options that serve as defaults, the command allows to append additional CMake options to the command-line. Options passed after a double-dash (--) are directly passed along. (For this the Spack main.py needed to be changed to allow this kind of unparsed command-line options)

I see this as highly valuable for the development process so you don't have to add new variants for new flags or re-concretize for every change. However, this definitely feels like a "use at your own risk" type of feature that would be powerful for application developers but that could also lead to confusion in delta's from the pure spack build for those who aren't careful.   

Also, I am not sure how I feel about the `--` interface.  A file interface would also be nice.
This is another case where I think generating an artifact to reproduce the build based on what you put in the CLI would be beneficial. 

I imagine whatever flags developers used here would/could also be used to drive what actually becomes a variant in the products spack package so a path to translating back to spack would be really great.

**Comment by knoepfel** [comment]
@rbberger, we at Fermilab have developed https://github.com/knoepfel/spack-mpd, which seems related to what you're working on.  I know @psakievich is already familiar with `spack-mpd`, but I refer to it here in case it would be beneficial to put our heads together.

**Comment by rbberger** [comment]
@psakievich Thank you for the detailed review. I'm all for making the existing tools better and maybe we can even avoid a whole new command while improving the developer experience. Sorry for the wall of text below :grin:

> It would be really nice to remove this as an assumption and make it a property of the command. For example, since we are assuming manipulation of the current shell, is there a compelling reason to not do `source $(spack build-env [package])` in the current shell? 

You probably meant something like `source <(spack build-env [package])`. There's currently two problems with that specific approach:

1. bad quoting of the some variables, like `SPACK_SYSTEM_DIRS`, `SPACK_COMPILER_FLAGS_REPLACE`
2. the missing export to make those variables active.

Both seem to be fixed when using the `--dump` option. So one could do it with multiple commands and a temporary file. That will likely be the most portable accross different shells. In any case, if this would happen within the tool execution it would actually make it more like `spack dev-build` and might replace the need for a new command.

> > This new command is closely related to spack develop and spack dev-build. However, unlike these two existing commands the goal of spack cmake is not to build and install the package within Spack, but allow it to be configured and built outside of the Spack build environment and within the same shell (no sub-shells)
> 
> Same notes as above, but also makes me wonder what else don't you like about the current tools? This command seems like one of the driving reasons is to move out of the current `spack` tools and back to what others are more used to. While there is merit to that, it would be helpful to understand what the deficiencies are since the spack tooling is generalized enough to support more build systems than just CMake and working on more than one application at at time.

The less steps you have to teach someone, the better. I see that the `spack develop` and `spack dev-build` approaches are too complicated to teach new code team members. There is just too many different commands and options flying around.

From my point of view, the main deficiencies of `spack dev-build` right now are:
- there is no way of defining a build directory via options, it has to be done indirectly, e.g. by create an environment and setting `package_attributes:build_directory` for that package.
- if a dependency isn't installed, it'll show all output of the dependency build as well. This can be worked around by first doing `spack install --only dependencies`.
- the default is to build **and install**. That case is already covered by `spack develop` and `spack install`. As a code developer, when you are working on the code in a particular configuration, the main focus is to build the current code, run unit tests, and continue working with that build directory. The install may only later become important.
- If you decide to test the install capabilities, you also don't want to install **into** Spack itself, but to some custom local prefix. That's one example of wanting to modify the CMake command-line (granted, it is easy to fix once you have a build folder or using `DEST`).
- the verbosity of the (CMake) output that can not be changed. It defaults to verbose Makefiles that are over the top for the common build. (see https://github.com/spack/spack/issues/43383)

Below is the same common developer / CI workflow using the different commands:

(Note I created these examples for simplicity, the actual environments we use don't use
`spack add`, but instead we create them by using a `spack.yaml` that has includes of common configuration)

### spack cmake

```bash
# create an environment and install the dependencies
spack env create lammps_env
spack env activate lammps_env
spack add lammps@develop~openmp
spack install --only dependencies

# configure project with cmake, inherit config from env, and activate testing
spack cmake -i -t -B build_serial -S . lammps

# build
cmake --build build_serial --parallel
```

### spack develop
```bash
# create an environment so we can use spack develop
spack env create lammps_env
spack env activate lammps_env
spack add lammps@develop~openmp
spack develop -b build_serial -p . --no-clone lammps@develop

spack install --test root -u cmake -v lammps
spack build-env --dump myenv.sh lammps
source myenv.sh

# build, without the noise
cmake -DCMAKE_VERBOSE_MAKEFILE=off build_serial
cmake --build build_serial --parallel
```

### spack dev-build
In this particular case, `dev-build` doesn't really add a benefit over `spack develop` + `spack install`
```bash
# workaround to set build directory, create an empty env and set packages_attributes:build_directory
spack env create lammps_env
spack env activate lammps_env
spack config add packages:lammps:package_attributes:build_directory:build_serial
spack dev-build --test root -u cmake lammps@develop~openmp

spack build-env --dump myenv.sh lammps
source myenv.sh

# build, without the noise
cmake -DCMAKE_VERBOSE_MAKEFILE=off build_serial
cmake --build build_serial --parallel
```

So as we can see, it's possible, but it's not great. It almost feels like `spack dev-build` is a great command that is waiting to come out of its shell, no pun intended.

> The only reason we launch subshells for the current commands is to keep the active shell unmodified. This could be reconfigured with an option to help with this CI use case without adding a whole new command/interface.

That by itself would already be an improvement and help make the case for `spack dev-build`. The main motivation is to have both developers and CI use the same commands to get to the same result. If I can get things done interactively with `dev-build`, but must hack around it when in a non-interactive context, it loses part of its appeal.

Maybe the answer is to make `spack dev-build` more like what `spack cmake` intends to do (with the proper build env), and avoid the duplication with `spack develop`. Just brainstorming here:  What would be great is if we could do something like this (which you might even simplify with an alias):

```bash
spack dev-build --env --test root -B build_serial -u cmake lammps@develop~openmp
cmake -B build_serial --parallel
```

> I see this as highly valuable for the development process so you don't have to add new variants for new flags or re-concretize for every change. However, this definitely feels like a "use at your own risk" type of feature that would be powerful for application developers but that could also lead to confusion in delta's from the pure spack build for those who aren't careful.

Removing the Spack wrappers / environment "magic" isn't an absolute must, it just makes it more approachable by devs that don't understand Spack in detail. But I agree, it comes at a cost and "you better know what you are doing" when you do this. So I do like having the option to use the build-env by default as long as it happens within the same command and manipulates the same shell.

Yes, avoiding the reconcretization and reusing the environment is a bit of a gimmick. But it does save people from having to manipulate the environment `spack.yaml` (which is yet another skill to learn) and write `spack concretize -f --fresh`.

**Comment by psakievich** [comment]
@rbberger I don't mind the wall of text.  I think your critiques of `spack dev-build` are valid and it would be good to make the improvements you've outlined.  I think your list of shortcomings are addressable in that command.  I do see a lot of value in reducing the entry point of developer workflows for people focused on app development.  

I created a different orchestration to help with this via https://github.com/sandialabs/spack-manager

Our workflow would look like this:

``` console
$ quick-create-dev --specs lammps@develop~openmp # orchestrates creation/activation/spack develop etc
$ spack install --test root -u cmake -v lammps
$ build-env-dive lammps # dives into a subshell with the build env active inside the build directory
# --- standard developer workflow
$ [code change]
$ make -j16
$ make test
```

In a nutshell, our approach is to have modular environment creation with shell commands to wrap multiple redundant steps together (`spack env create/activate/develop`, `source spack-build-env.txt && spack cd -b [product]`).

Granted this does use a subshell which is on your wish-list to remove but it has worked well for our developers.  In the projects I've worked on the trade off is that developers are frequently modifying multiple parts of the graph with multiple develop specs so a subshell per product is actually quite nice.

I can see how for a single application this flexibility might not be required which is where `spack dev-build` could shine with a little more polishing.  I would advise we put our efforts there first and see what is left once PR's are reviewed and merged. If the only remaining critique after that is the spack commands are a little verbose/option heavy (perfectly valid critique) then that seems like a ripe opportunity to employ spack's aliases and/or a shell orchestration function.

 @tgamblin that's my 2x :coin:

**Comment by junghans** [comment]
I like this approach! 

+1 from my side.

**Comment by biddisco** [comment]
I pushed an updated version of this branch with one fix for the builder access - the following line causes an error on latest develop
```
else [x for x in app.builder.std_cmake_args if any(n in x for n in accepted_std_args)]
```
I rebased onto develop, but I cannot open a PR to merge into this PR for some reason. The updated branch is here 
https://github.com/biddisco/spack/tree/spack_cmake_cmd

**Comment by rbberger** [comment]
@biddisco thanks for the update. I pushed your commit and redid the rebase. Not sure if this MR is going anywhere. I think the consensus was that it'll be better to improve the existing tools. Nevertheless, if people find this useful it's ok to keep it working.

**Comment by biddisco** [comment]
> @biddisco thanks for the update. I pushed your commit and redid the rebase. Not sure if this MR is going anywhere. I think the consensus was that it'll be better to improve the existing tools. Nevertheless, if people find this useful it's ok to keep it working.

I find it incredibly useful and use it constantly - however, I was using the stand-alone `spack_cmake` tool that you first created outside of the spack source tree and when it broke yesterday I thought I'd better try the version in the PR. I don't know if it is easier for people like me to keep it maintained in the source tree or out of it. Keeping branches synced/merged with latest develop is not a big deal - until something major changes, and then I'm stuffed! So if there is anything we can do to persuade the spack masters to accept the contribution and maintain it, then I will do what I can. To that end, I'll read the comments in this PR thoroughly and see if I can do anything useful to move it along.

**Comment by rbberger** [comment]
Closing, since we're going in a different direction.

**Comment by biddisco** [comment]
I still use this on a daily basis after keeping it update. Is there something that will replace it?

**Comment by rbberger** [comment]
@biddisco one of the outcomes of the Spack User Meeting in Chicago last month was a general consensus that we're all kind of solving the same issues and we should collaborate more on finding better solutions for this. Based on the energy at that conference, I'm hopeful you will see something come out of it in the near future. But I see this PR as a dead end.

In the meantime, my current recommendation is going the `spack develop` route along with `spack build-env`. [I've presented](https://www.youtube.com/watch?v=u73mESk0Q4E&list=PLRKq_yxxHw29-JcpG2CZ-xKK2U8Hw8O1t&index=33) my current workflow that doesn't use "spack cmake" anymore, but I encourage you to look at other talks of that conference as well.
